crates/services/txpool_v2/src/storage/mod.rs

```diff
@@ -36,34 +38,53 @@ pub struct StorageData {
 
 pub type RemovedTransactions = Vec<PoolTransaction>;
 
+pub trait Collision<StorageIndex> {
+    /// Returns the instigator of the collision.
+    fn tx(&self) -> &PoolTransaction;
+
+    /// Returns the transactions that collide with the given transaction.
+    fn colliding_transactions(&self) -> &HashMap<StorageIndex, Vec<CollisionReason>>;
+
+    /// Unwraps the collision into a instigator transaction.
+    fn into_instigator(self) -> PoolTransaction;
+}
+
+pub trait CheckedCollision<StorageIndex>: Collision<StorageIndex> {
+    /// Returns the list of all dependencies of the transaction.
+    fn all_dependencies(&self) -> &HashSet<StorageIndex>;
+}
+
 /// The storage trait is responsible for storing transactions.
 /// It has to manage somehow the dependencies between transactions.
 pub trait Storage {
     /// The index type used in the storage and allow other components to reference transactions.
-    type StorageIndex: Copy + Debug;
+    type StorageIndex: Copy + Debug + Eq + Hash;
+    type CheckedCollision<C>: CheckedCollision<Self::StorageIndex>
+    where
+        C: Collision<Self::StorageIndex>;
 
     /// Store a transaction in the storage according to the dependencies.
     /// Returns the index of the stored transaction.
-    fn store_transaction(
+    fn store_transaction<C>(
         &mut self,
-        transaction: PoolTransaction,
+        checked_collision: Self::CheckedCollision<C>,
         creation_instant: Instant,
-    ) -> Result<Self::StorageIndex, Error>;
+    ) -> Self::StorageIndex
+    where
+        C: Collision<Self::StorageIndex>;
 
     /// Check if a transaction could be stored in the storage. This shouldn't be expected to be called before store_transaction.
     /// Its just a way to perform some checks without storing the transaction.
-    fn can_store_transaction(
+    fn can_store_transaction<C>(
         &self,
-        transaction: &PoolTransaction,
-        collisions: Option<Vec<Self::StorageIndex>>,
-    ) -> Result<(), Error>;
+        collision: C,
```

Comment:
Same comment has above not having a `transaction` parameter here is very strange when reading the code.

Reply:
It is just an explicit representation of the way you choose before=) Before, you required transactions and a list of collisions as input parameters. Now, it is represented by one type. We can rename it to be `TransactionWithCollisions`, and it will fit better. But the main question here is, should `Storage` be responsible for verifying the transaction?=)